Source/DafnyLanguageServer.Test/Util/TestNotificationReceiver.cs

```diff
@@ -26,46 +25,43 @@ public class TestNotificationReceiver<TNotification> {
     }
 
     public void NotificationReceived(TNotification request) {
-      logger.LogTrace($"Received {request.Stringify()}");
       Assert.NotNull(request);
-      notifications.Enqueue(request);
-      notificationHistory.Add(request);
-      availableNotifications.Release();
+      logger.LogTrace($"Received {request.Stringify()}");
+
+      lock (this) {
+        notificationHistory.Add(request);
+        notifications.Enqueue(request);
+      }
     }
 
-    public bool HasPendingNotifications => !notifications.IsEmpty;
+    public bool HasPendingNotifications => notifications.Size != 0;
 
     public IReadOnlyList<TNotification> History => notificationHistory;
 
     public void ClearHistory() {
       notificationHistory.Clear();
     }
 
-    public TNotification GetLast(Func<TNotification, bool> predicate) {
-      var result = History.LastOrDefault(predicate);
-      ClearQueue();
-      return result;
-    }
+    public TNotification? GetLatestAndClearQueue(Func<TNotification, bool> predicate) {
+      lock (this) {
```

Comment:
Oh, great, that way the history and the notification queue are in sync. Yes, that is definitely something necessary.
Note that, if you lock this, you can now reuse `clearQueue()` as before, it should not matter and would be more readable.

The implicit precondition of this method is that the notifications queue is a suffix of the history.

Reply:
`AsyncQueue` does not have a `Clear` method, so I have to do 

```c#
while (notifications.Size > 0) {
  _ = notifications.Dequeue(CancellationToken.None);
}
```

to clear it.

> The implicit precondition of this method is that the notifications queue is a suffix of the history.

Once you enter the lock, yes.